### DISCUSSION TITLE
Versioning

### DISCUSSION THREAD

**Discussion by borkdude** [clj-easy/tools.misc#2] [Ideas]
Maybe it's a good idea to version the protocols:

```
HttpClient_v1
```
so we can add more protocols, `HttpClient_v2`, over time. Old libraries keep working if they encounter an implementation of v1, but can upgrade to the improved protocol v2, by using that.

Alternatively, one could add more methods to the protocol over time.

We can start with `v0` which is the alpha and allow breakage on that one.

**Comment by mynomoto** [comment]
You could use the namespace for versioning the protocol. Maybe `clj-easy.tools.http-client.v0`?

**Reply by borkdude** [reply]
That might be even better since you won't have clashes between var names then.

**Reply by borkdude** [reply]
I think it could even make sense to version on the function level:

```
(defn write-str-v1 [s key-fn])
(defn write-str-v2 [s key-fn new-option])
```

This way we can deal with new versions of functions without having an implementor create an additional namespace and re-implement all of the functions when only some of them have changed.

**Reply by jeroenvandijk** [reply]
_For my own understanding I'm writing how I understand this. Hopefully it is useful for you as well._

I'm thinking there are basically 3 scenario's possible

### Interface and provider versions are equal
Everything good

### Old interface, new provider
Two options:

1. Interface needs to be updated
2.  An older version of the provider (packaged with the new provider) should be used

Option 2 would not work with protocols (because one of the extended protocols would be non-existent), but this would work if providers add versioned keys in a map (see below). However, option 1 seems to be ok too and in control of the user. Option 2 is still preferable as it would require less work and understanding of the end user.

### New interface, old provider
I'm not sure about this one, but I'm guessing this would require the interface library to try to call the old interface after failed attempts to find a new implementation.

This could be done with both protocols and the map approach

Protocols:
```clojure
(extend-type Object
   V2Write
   (write-str-v2 [parser s]
      (write-str-v1 parser (assoc s :new-key :value)))
``` 

Map approach seems to be cleaner as it would allow us to do this conversion on registration.
```clojure
(defn reg-provider!  [id opts]
  (if (compatible-v2? opts)
     (vswap! provider-impls assoc id opts)
     (vswap! provider-impls assoc id (add-v2-options opts))))
```
 Note that if the provider is already at version 3 the map approach could also work without issues.

**Reply by borkdude** [reply]
> Old interface, new provider

I would say that new providers should always add backward compatibility with the old interfaces too, so callers don't get broken when they upgrade the provider. @mynomoto's approach can be used by versioning the namespaces like `clj-easy.tools.json.v1.clj`, `clj-easy.tools.json.v2.clj`, etc.

**Reply by borkdude** [reply]
I think using the namespace approach also takes care of @bsless's objection against using volatile indirections. The presence of the namespace on the classpath is the interface implementation. By default, libraries can program against an empty default implementation:

```
(ns clj-easy.tools.json.v0)

(defn read-str [...])
(defn write-str [...])
```

maybe deployed as `[clj-easy/tools.json "0.0.1"]` or whatever.

Then users can pull in a cheshire (or whatever) implementation by using `[clj-easy/tools.json "0.0.1-cheshire"]`. But how does a dependency tool favor that over the empty library? Must `:exclusions` be used?

**Reply by mynomoto** [reply]
Dependencies on the top level win so if the user declares the dependency it will work. I think there should not be an empty version because that will cause things to not work at all.

**Reply by borkdude** [reply]
> I think there should not be an empty version because that will cause things to not work at all.

I think there should, as libraries should be able to program against the interface, not against an implementation. How else are you going to do this?

**Reply by borkdude** [reply]
Another example I have in mind: I currently have ASM as a dependency of clj-kondo to do Java analysis. But not everybody is interested in Java analysis and the ASM dependency may be problematic in combination with other libs. So I want to be able to provide a no-op implementation by default for Java analysis and a library which overrides the namespace with an ASM implementation must be used to get java analysis.

**Reply by mynomoto** [reply]
> I think there should, as libraries should be able to program against the interface, not against an implementation. How else are you going to do this?

I think if the implementation used by the library author then this is not useful. The library author should be able to choose any implementation and the user could change it and the library should work. Else this is not swapable right?

**Reply by borkdude** [reply]
clojure.tools.logging does not provide any logging implementation: the user should bring this. That's the point, a separation between the interface and the implementations. No implementations on the classpath by default.

**Reply by borkdude** [reply]
if you're wondering: but how can libraries test their features that depend on an implementation? Well, they declare an implementation in their test dependencies :) 
But maybe you're right, and it would be more convenient if you had one implementation by default. Maybe it doesn't matter that much :). (As you can see, I'm often thinking out loud and change my mind after a while, that's what these discussions are good for right?)

**Reply by mynomoto** [reply]
> As you can see, I'm often thinking out loud and change my mind after a while, that's what these discussions are good for right?

Yeah, I'm surely enjoying this. It is all about choices and there is no right choice that will work all the time for everyone.

**Reply by borkdude** [reply]
I can see situations where bringing no implementation can be useful, especially with library A depending on library B and B uses YAML but only in some optional part which is non-essential and library A depends on B but doesn't want to have YAML in its transitive dependency tree, since YAML has nothing to do with what library A is about. Then B could use the empty YAML provider (which maybe throws on first usage).

**Reply by mynomoto** [reply]
Oh, that's interesting. I'm the end, libraries don't know their transitive dependencies because users are always able to override.
I would be concerned by codepaths that are less commonly used having those dependencies and it blows up in prod but people also like less dependencies.

**Reply by borkdude** [reply]
Yeah. So, a library could default to one implementation, but an app or other library could decide to pull in `clj-easy/tools.json {:mvn/version "0.1.0-none"}`.  I wonder what happens if library B uses "0.1.0-cheshire" and library A uses "0.1.0-none" (since it's pretty sure nobody wants to use YAML) and an app uses library A (and/or B).

**Reply by mynomoto** [reply]
If the app has no top level dependency this will depend on tooling because tools.deps and maven have different ways to pick transitive dependencies. Iirc tools.deps tries to pick more recent versions and maven is dependent on the order of deps.

**Reply by jeroenvandijk** [reply]
Let's say some library depends on the `clj-easy/tools.json` interface, using the above method, they there are two options of including the library

1. declaring the dependency as provided "provided" or not declaring it at all. The end user has to include the dependency.
2. using a `clj-easy/tools.json` version "none" that throws on usage. The end user has to override this dependency with a version that works as desired

I see **downsides** with both approaches:

1. Will blow up at runtime leading to non-intuitive missing namespace errors ([reminding me of ring](https://github.com/ring-clojure/ring#upgrade-notice))
2. Will not work perfectly with 3rd party extensions as 3rd parties will not be able to create an artifact on clojars with `clj-easy/tools.json`. This would only work for git repo's in combination with `:git/url`

Potential **solution** to problem 2:

Using a macro that resolves to the latest available function and has a default implementation that throws. WDYT? Code below

```clojure

;; Example implementation
(ns clj-easy.tools.json.v3)

(defn read-str [& _]
  "v3 result here")

;; Main interface
(ns clj-easy.tools.json)

(defmacro def-fn [fqn-sym latest-version dep]
  (let [ns-name (symbol (namespace fqn-sym))
        fn-sym (symbol (name fqn-sym))
        f (some
           (fn [version]
             (when-let [n (find-ns (symbol (str ns-name ".v" version)))]
               (ns-resolve n fn-sym)))
           (reverse (range 0 (inc latest-version))))]
    `(def ~fn-sym
       (or ~f
           (fn [& args#]
             (throw
              (ex-info (str "No implementation for '" (quote ~ns-name) "/" (quote ~fn-sym) ". "
                            "Please add a " ~dep  " implementation "
                            "to the dependencies.") {})))))))

(def-fn clj-easy.tools.json/read-str 3 "clj-easy/tools.json")

(read-str "hi") ;=> "v3 result here"

(def-fn clj-easy.tools.json/read-str 2 "clj-easy/tools.json")

(read-str "hi") ;=> error


```

**Reply by borkdude** [reply]
> Will not work perfectly with 3rd party extensions as 3rd parties will not be able to create an artifact on clojars with clj-easy/tools.json.

This might be a reason why we should then centralize implementations, like @mynomoto has suggested.
But you could also have a global exclusion on `clj-easy.tools.json` in your deps.edn and then pull in a 3rd party package.

**Comment by borkdude** [comment]
@mynomoto I made a branch, aptly named `mynomoto-solution` [here](https://github.com/clj-easy/tools.misc/tree/mynomoto-solution).

I think I did what you suggested there.

When you run `bb test:clj` the test runner is invoked twice but with a different classpath, providing implementations for cheshire and data.json.

This is the cheshire implementation:

https://github.com/clj-easy/tools.misc/blob/mynomoto-solution/json/cheshire/clj_easy/tools/json/v0.clj

**Reply by mynomoto** [reply]
Yeah, that is what I was thinking. It looks great and it is way simpler than what I have on poly so far. I like your version better.
I would add a couple of tests reading a malformed json and writing some java object. Then I think we will start to see if this idea will be practical or not.

**Reply by borkdude** [reply]
I merged to master. PR for those tests welcome. Happy we reached some kind of consensus.

**Comment by borkdude** [comment]
A demo of the current approach on main, using cheshire as the json impl:

```clojure
(require '[babashka.deps])
(babashka.deps/add-deps '{:deps {io.clj-easy.tools.json
                                 {:git/url "https://github.com/clj-easy/tools.misc"
                                  :deps/root "json/cheshire"
                                  :git/sha "aea9d3165650b82de4332c9ae53fb5f0f01a5ce0"}}})

(require '[clj-easy.tools.json.v0 :as json])
(prn (json/read-str (json/write-str {:a 1})))
```

**Reply by borkdude** [reply]
LOL, but also the data.json one works, because bb can execute that from source:

``` clojure
(require '[babashka.deps])
(babashka.deps/add-deps '{:deps {io.clj-easy.tools.json
                                 {:git/url "https://github.com/clj-easy/tools.misc"
                                  :deps/root "json/data-json"
                                  :git/sha "aea9d3165650b82de4332c9ae53fb5f0f01a5ce0"}}})

(require '[clj-easy.tools.json.v0 :as json])
(prn (json/read-str (json/write-str {:a 1})))
```
